{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../popalign/')\n",
    "import popalign as PA\n",
    "import importlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path of the popalign library\n",
    "PA.__file__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data\n",
    "Run the first cell if samples are stored individually  \n",
    "Run the second cell if samples are stored in the same matrix (requires meta data file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Load samples from distinct matrix files\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "samples : dict\n",
    "    Dictionary of sample names (keys) and paths to their respective matrix files (values)\n",
    "genes : str\n",
    "    Path to a .tsv 10X gene file. Optional if existing_obj is provided\n",
    "outputfolder : str\n",
    "    Path (or name) of the output folder to create\n",
    "existing_obj : dict, optional\n",
    "    Object previously returned by either load_samples() or load_screen(). New samples will be added to that object\n",
    "'''\n",
    "\n",
    "mysamples = {\n",
    "    'CTRL' : '../data/samples/PBMC.mtx',\n",
    "    'GMCSF_1ng/ml' : '../data/samples/GMCSF.mtx',\n",
    "    'IFNG_1ng/ml' : '../data/samples/IFNG.mtx',\n",
    "    'IL2_10ng/ml' : '../data/samples/IL2.mtx',\n",
    "    'CD40L_20ng/ml' : '../data/samples/CD40L.mtx',\n",
    "}\n",
    "mygenes = '../data/samples/genes.tsv'\n",
    "\n",
    "pop = PA.load_samples(samples=mysamples, \n",
    "                      genes=mygenes,\n",
    "                      outputfolder='output_samples',\n",
    "                      existing_obj=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Load data from a multiplexed experiment\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "matrix : str\n",
    "    Path to a sparse matrix\n",
    "barcodes : str\n",
    "    Path to a .tsv 10X barcodes file\n",
    "metafile : str\n",
    "    Path to a metadata file. Must contains `cell_barcodes` and `sample_id` columns\n",
    "genes : str\n",
    "    Path to a .tsv 10X gene file. Optional if existing_obj is provided\n",
    "outputfolder : str, optional\n",
    "    Path (or name) of the output folder to create\n",
    "existing_obj : dict, optional\n",
    "    Object previously returned by either load_samples() or load_screen(). New samples will be added to that object\n",
    "only : list, optional\n",
    "    List of sample names to load (other samples with names not in list will not be loaded)\n",
    "col : str, optional\n",
    "    Name of a specific column in the meta data to use. Defaults to None\n",
    "value : str or int, optional\n",
    "    Value in the specified meta data column `col` to use to filter samples to load. Defaults to None\n",
    "'''\n",
    "mymatrix = '../data/screen/PBMC-MULT-6/PBMC-MULT-6.mtx'\n",
    "mybarcodes = '../data/screen/PBMC-MULT-6/barcodes.tsv'\n",
    "mygenes = '../data/screen/PBMC-MULT-6/genes.tsv'\n",
    "mymetadata = '../data/screen/PBMC-MULT-6/updated_meta.csv'\n",
    "\n",
    "pop = PA.load_multiplexed(matrix=mymatrix, \n",
    "                     barcodes=mybarcodes,\n",
    "                     genes=mygenes,\n",
    "                     metafile=mymetadata,\n",
    "                     outputfolder='output_pbmc6_matt',\n",
    "                     only=[], # list of sample names to only load the specified samples\n",
    "                     col=None, # either None or a column name from the meta data\n",
    "                     value=None,\n",
    "                     existing_obj=None) # if col != None, specify value in column to filter samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "PICKLE TO SAVE THE DATA\n",
    "LOAD THE DATA VERY EFFICIENTLY\n",
    "'''\n",
    "\n",
    "import pickle\n",
    "with open('mult6_data.pickle', 'wb') as handle:\n",
    "    pickle.dump(pop, handle)\n",
    "\n",
    "#with open('mult6_data.pickle', 'rb') as handle:\n",
    "#    pop = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Display number of cells in each sample, total number of cells\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "'''\n",
    "PA.print_ncells(pop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize and gene filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Normalize the samples of object `pop` and applies a normalization factor\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "scaling_factor : int or None, optional\n",
    "    Number used to scale the data values. If None, that factor is computed automatically\n",
    "ncells : int or None\n",
    "    Number of cells to randomly subsample to try different normalization factors to use less memory. If None, all cells are used.\n",
    "'''\n",
    "\n",
    "PA.normalize(pop, scaling_factor=None, ncells=30000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Plot genes by their log(mean) and log(coefficient of variation)\n",
    "Can be used multiple times to find best offset\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "offset: float\n",
    "    Value (its log) will be added to the intercept of the linear fit to filter genes\n",
    "'''\n",
    "\n",
    "PA.plot_gene_filter(pop, offset=1.13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Filter genes from data in `pop`\n",
    "Can discard ribosomal genes that start with RPS or RPL\n",
    "Can discard mitochondrial genes that start with MT-\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop :dict\n",
    "    Popalign object\n",
    "remove_ribsomal : bool\n",
    "    Wether to remove or not the ribosomal genes\n",
    "remove_mitochondrial : bool\n",
    "    Wether to remove or not the mitochondrial geneset\n",
    "'''\n",
    "PA.filter(pop, remove_ribsomal=True, remove_mitochondrial=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Remove red blood cells from data\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "species : str\n",
    "    name of experiment species. Can be human or mouse\n",
    "'''\n",
    "#PA.removeRBC(pop, species='human')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dimensionality reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Compute feature spaces and minimize the reconstruction error to pick a final feature space\n",
    "Run Gene Set Enrichment Analysis (GSEA)\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "ncells : int\n",
    "    Number of cells to use\n",
    "nfeats : int or list of ints\n",
    "    Number(s) of features to use\n",
    "nreps : int\n",
    "    Number of repetitions to perform for each k in nfeats\n",
    "niter : int\n",
    "    Maximum number of iterations to perform for each instance of the algoirthm\n",
    "'''\n",
    "\n",
    "importlib.reload(PA)\n",
    "PA.onmf(pop, ncells=5000, nfeats=[7,9], nreps=3, niter=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grid of samples in embedding space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate a grid plot of sample plots in an embedding space\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "method : str\n",
    "    Embedding method. One of tsne, umap\n",
    "figsize : tuple\n",
    "    Figure size\n",
    "size_background : float, int\n",
    "    Point size for the embedding scatter in the background\n",
    "size_samples : float, int\n",
    "    Point size for the highlighted samples\n",
    "'''\n",
    "\n",
    "PA.samples_grid(pop, method='tsne', figsize=(20,20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build probabilistic models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Build a Gaussian Mixture Model on feature projected data for each sample\n",
    "Type model populations\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "ks : int or tuple\n",
    "    Number or range of components to use\n",
    "niters : int\n",
    "    number of replicates to build for each k in `ks`\n",
    "training : int or float\n",
    "    If training is float, the value will be used a percentage to select cells for the training set. Must follow 0<value<1\n",
    "    If training is int, that number of cells will be used for the training set.\n",
    "nreplicates : int\n",
    "    Number of replicates to generate. These replicates model will be used to provide confidence intervals later in the analysis.\n",
    "reg_covar : str or float\n",
    "    If 'auto', the regularization value will be computed from the feature data\n",
    "    If float, value will be used as reg_covar parameter to build GMMs\n",
    "rendering : str\n",
    "    One of grouped, individual\n",
    "types : dict, str or None\n",
    "    Dictionary of cell types. Keys should be cell types labels and values should be lists of valid genes\n",
    "    If None, a default PBMC cell types dictionary is provided\n",
    "figsizegrouped : tuple, optional\n",
    "    Size of the figure for the renderings together. Default is (20,20)\n",
    "figsizesingle : tuple, optional\n",
    "    Size of the figure for each single sample rendering. Default is (5,5)\n",
    "only: list or str, optional\n",
    "    Sample label or list of sample labels. Will force GMM construction for specified samples only. Defaults to None\n",
    "'''\n",
    "\n",
    "importlib.reload(PA)\n",
    "PA.build_gmms(pop, \n",
    "              ks=(3,12), \n",
    "              niters=2,\n",
    "              training=.8, \n",
    "              nreplicates=0,\n",
    "              reg_covar='auto', \n",
    "              rendering='grouped', \n",
    "              types='defaultpbmc', # either None, 'defaultpbmc' or a dictionary\n",
    "              figsizegrouped=(20,20),\n",
    "              figsizesingle=(6,5),\n",
    "              only=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If the renderings do not have the right figure size, regenerate the renderings with:\n",
    "'''\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "figsizegrouped : tuple\n",
    "    Figure size for the grid rendering plotof all samples together\n",
    "figsizesingle : tuple\n",
    "    Figure size of an individual sample rendering plot\n",
    "mode : str\n",
    "    One of grouped, individual or unique.\n",
    "    Grouped will render the models individually and together in a separate grid\n",
    "    Inidividual will only render the models individually\n",
    "    Unique will render the data's unique model\n",
    "'''\n",
    "importlib.reload(PA)\n",
    "PA.render_models(pop, figsizegrouped=(30,30), samples=order, figsizesingle=(6,5), mode='grouped')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align subpopulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Align the components of each sample's model to the components of a reference model\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "ref : str\n",
    "    Name of reference sample\n",
    "method : str\n",
    "    Method to perform the alignment\n",
    "    If 'conservative', the reference component and the test component have to be each other's best match to align\n",
    "    If 'test2ref', the closest reference component is found for each test component\n",
    "    If 'ref2test', the closest test component is found for each test component\n",
    "figsizedeltas : tuple, optional\n",
    "    Size of the figure for the delta plot. Default is (10,5)\n",
    "figsizeentropy : tuple, optional\n",
    "    Size of the figure for the entropy plot. Default is (10,5)\n",
    "'''\n",
    "importlib.reload(PA)\n",
    "PA.align(pop, ref='CONTROL_CD3',\n",
    "         method='test2ref', # one of: test2ref, ref2test, conservative\n",
    "         figsizedeltas=(10,10),\n",
    "         figsizeentropy=(10,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(PA)\n",
    "PA.plot_deltas_test(pop, 20, (15,15)) # pop, pointsize, figsize arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "SAVE ALIGNMENT MATRICES TO CSV FILES\n",
    "'''\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "dname = 'alignment_matrices_control_cd3'\n",
    "PA.mkdir(os.path.join(pop['output'], dname))\n",
    "\n",
    "cols = ['%d %s' % (i,s) for i,s in enumerate(pop['samples']['CONTROL_CD3']['gmm_types'])]\n",
    "for x in pop['order']:\n",
    "    try:\n",
    "        arr = pop['samples'][x]['fullalignments']\n",
    "        rows = ['%d %s' % (i,s) for i,s in enumerate(pop['samples'][x]['gmm_types'])]\n",
    "        df = pd.DataFrame(arr, index=rows, columns=cols)\n",
    "        df.to_csv(os.path.join(pop['output'], dname, '%s.csv' % x))\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Find  differentially expressed genes between a refernce subpopulation\n",
    "and the subpopulation of a sample that aligned to it\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "refcomp : int\n",
    "    Subpopulation number of the reference sample's GMM\n",
    "sample : str\n",
    "    Name of the sample to compare\n",
    "nbins : int, optional\n",
    "    Number of histogram bins to use\n",
    "nleft : int\n",
    "    Number of underexpressed genes to retrieve\n",
    "nright : int\n",
    "    Number of overexpressed genes to retrieve\n",
    "renderhists : bool\n",
    "    Render histograms or not for the top differentially expressed genes\n",
    "usefiltered : bool\n",
    "    Wether to use filtered genes or not. If False, all genes will be used to run the differential expression\n",
    "'''\n",
    "\n",
    "refcomp = 1\n",
    "sample = 'Budesonide'\n",
    "genelist = PA.diffexp(pop,\n",
    "                      refcomp=refcomp,\n",
    "                      sample=sample,\n",
    "                      nbins=20,\n",
    "                      cutoff=.6,\n",
    "                      renderhists=True,\n",
    "                      usefiltered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Plot specific genes for cells of a reference subpopulation S and subpopulations that aligned to S\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign dict\n",
    "refcomp : int\n",
    "    Reference subpopulation number\n",
    "genelist : list\n",
    "    List of genes to plot\n",
    "clustersamples : bool\n",
    "    Cluster the samples\n",
    "clustercells : bool\n",
    "    Cluster the cells within each sample subpopulation\n",
    "savename : str, optional\n",
    "    The user can specify a name for the file to be written. When savename is None, a filename is computed with the reference component number. Default is None\n",
    "figsize : tuple, optional\n",
    "    Size of the figure. Default is (15,15)\n",
    "cmap : str, optional\n",
    "    Name of the Matplotlib colormap to use. Default is Purples\n",
    "samplelimits : bool, optional\n",
    "    Wether to draw vertical lines on the heatmap to visually separate cells from different samples\n",
    "scalegenes : bool, optional\n",
    "    Wether to scale the genes by substracting the min and dividing by the max for each gene\n",
    "only: str, optional\n",
    "    Sample label\n",
    "equalncells: bool, optional\n",
    "    Randomly subsets cells so that subpopulations have the same number of cells\n",
    "'''\n",
    "\n",
    "PA.plot_heatmap(pop,\n",
    "                refcomp=refcomp,\n",
    "                genelist=genelist,\n",
    "                clustersamples=True,\n",
    "                clustercells=False,\n",
    "                savename='%d_%s' % (refcomp,sample) ,# either None or a string\n",
    "                figsize=(20,20),\n",
    "                cmap='Purples',\n",
    "                samplelimits=False,\n",
    "                scalegenes=True,\n",
    "                only=None,\n",
    "                equalncells=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize genes in subpopulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Plot a heatmap of genes ~ GMM subpopulations cells for a given model\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "sample : str\n",
    "    Sample name to select model from pop dictionary\n",
    "genelist : list\n",
    "    List of gene names. If empty, the filtered genes will be used.\n",
    "savename : str\n",
    "    File name to use\n",
    "metric : str\n",
    "    Metric to use to cluster\n",
    "method : str\n",
    "    Method to use to cluster\n",
    "clustergenes : boolean\n",
    "    Wether or not to cluster the genes (rows)\n",
    "cmap : str\n",
    "    Name of the colormap to use\n",
    "figsize : tuple\n",
    "    Figure size\n",
    "'''\n",
    "PA.plot_genes_gmm_cells(pop,\n",
    "               sample='unique',\n",
    "               genelist=genelist,\n",
    "               savename='',\n",
    "               metric='correlation',\n",
    "               method='single',\n",
    "               clustergenes=False,\n",
    "               clustercells=False,\n",
    "               cmap='magma',\n",
    "               figsize=(15,15)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate a ranking plot of the samples against a reference model\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "ref : str\n",
    "    Reference sample name\n",
    "k : int\n",
    "    Number of random cells to use\n",
    "niter : int\n",
    "    Number of iterations to perform\n",
    "method : str\n",
    "    Scoring method to use. 'LLR' for log-likelihood ratio, 'LL' for log-likelihood.\n",
    "mincells : int\n",
    "    If a sample has less than `mincells` cells, is discarded\n",
    "figsize : tuple, optional\n",
    "    Size of the figure. Default is (10,5)\n",
    "'''\n",
    "\n",
    "PA.rank(pop,\n",
    "        ref='CONTROL', # label of the reference sample\n",
    "        k=100, # number of cells to randomly sample from sample\n",
    "        niter=200, # number of iterations\n",
    "        method='LLR', # LLR for log-likelihood ratio or LL for log-likelihood\n",
    "        mincells=50, # sample's minimum number of cells to be ranked\n",
    "        figsize=(10,5)) # plot figure size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build unique GMM for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Build a unique Gaussian Mixture Model on the feature projected data\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "ks : int or tuple\n",
    "    Number or range of components to use\n",
    "niters : int\n",
    "    number of replicates to build for each k in `ks`\n",
    "training : int or float\n",
    "    If training is float, the value will be used a percentage to select cells for the training set. Must follow 0<value<1\n",
    "    If training is int, that number of cells will be used for the training set.\n",
    "reg_covar : boolean or float\n",
    "    If True, the regularization value will be computed from the feature data\n",
    "    If False, 1e-6 default value is used\n",
    "    If float, value will be used as reg_covar parameter to build GMMs\n",
    "types : dict, str or None\n",
    "    Dictionary of cell types.\n",
    "    If None, a default PBMC cell types dictionary is provided\n",
    "figsize : tuple, optional\n",
    "    Size of the figure. Default is (6,5)\n",
    "'''\n",
    "PA.build_unique_gmm(pop, \n",
    "                    ks=(5,20), \n",
    "                    niters=3, \n",
    "                    training=0.4, \n",
    "                    reg_covar=True, \n",
    "                    types='defaultpbmc', # either None, 'defaultpbmc' or a dictionary\n",
    "                    figsize=(6,5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Plot proportions of the samples in each of the GMM components\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop: dict\n",
    "    Popalign object\n",
    "pcells: float\n",
    "    Percentage of random cells to use for each query repetition\n",
    "nreps: int\n",
    "    Number of times to repeat the query process with different cells for each sample\n",
    "figsize : tuple, optional\n",
    "    Size of the figure. Default is (5,20)\n",
    "'''\n",
    "PA.plot_query(pop,  pcells=.2, nreps=10, figsize=(10,20), sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2D visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Run an embedding algorithm and plot the data in a scatter plot\n",
    "Higlight specific gene expression\n",
    "\n",
    "pop: dict\n",
    "    Popalign object\n",
    "method : str\n",
    "    Embedding method. One of umap, tsne. Defaults to tsne\n",
    "sample: str, optional\n",
    "    Highlight specified sample\n",
    "compnumber: int, optional\n",
    "    Component number of specified sample\n",
    "marker : str\n",
    "    Valid gene name. Defaults to None\n",
    "size : float or int\n",
    "    Point size. Defaults to .1\n",
    "'''\n",
    "PA.scatter(pop, method='tsne', marker='CD3D')\n",
    "PA.scatter(pop, method='tsne', marker='LYZ')\n",
    "PA.scatter(pop, method='tsne', marker='CD163')\n",
    "PA.scatter(pop, method='tsne', sample='Budesonide', marker='CD163')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate a grid plot of sample plots in an embedding space\n",
    "\n",
    "Parameters\n",
    "----------\n",
    "pop : dict\n",
    "    Popalign object\n",
    "method : str\n",
    "    Embedding method. One of tsne, umap\n",
    "figsize : tuple\n",
    "    Figure size\n",
    "size_background : float, int\n",
    "    Point size for the embedding scatter in the background\n",
    "size_samples : float, int\n",
    "    Point size for the highlighted samples\n",
    "'''\n",
    "\n",
    "PA.samples_grid(pop, method='tsne', figsize=(20,20))\n",
    "PA.samples_grid(pop, method='umap', figsize=(20,20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate grid plots of sample subpopulations in an embedding space\n",
    "\n",
    "pop : dict\n",
    "    Popalign object\n",
    "method : str\n",
    "    Embedding method. One of tsne, umap\n",
    "figsize : tuple\n",
    "    Size of the figures to be generated\n",
    "size_background : float, int\n",
    "    Point size for the embedding scatter in the background\n",
    "size_subpops : float, int\n",
    "    Point size for the highlighted subpopulations\n",
    "'''\n",
    "PA.subpopulations_grid(pop, method='tsne', figsize=(20,20), size_subpops=20)\n",
    "#PA.subpopulations_grid(pop, method='umap', figsize=(20,20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3D visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Interactive 3D visualization of the data in feature space\n",
    "importlib.reload(PA)\n",
    "import plotly\n",
    "plotly.offline.init_notebook_mode()\n",
    "PA.plotfeatures(pop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Work in progress"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "#-------------- WORK IN PROGRESS BELOW ----------------\n",
    "#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n",
    "#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "def diffexp_two_matrices(M1, M2, genes, nbins=20, cutoff=.7):\n",
    "    M1 = M1.toarray()\n",
    "    M2 = M2.toarray()\n",
    "    \n",
    "    with Pool(pop['ncores']) as p:\n",
    "        q = np.array(p.starmap(PA.l1norm, [(ig, M1[ig,:], M2[ig,:], nbins) for ig in range(M1.shape[0])])) # for each gene idx ig, call the l1norm function\n",
    "    \n",
    "    idx = np.argsort(q)\n",
    "    q = q[idx]\n",
    "    genes = np.array(genes)\n",
    "    genes = genes[idx]\n",
    "    \n",
    "    downregulated_idx = np.where(np.array(q)<-cutoff)[0] # get indices of genes with low l1-norm values\n",
    "    upregulated_idx = np.where(np.array(q)>cutoff)[0] # get indices of genes with high l1-norm values\n",
    "    downregulated = [genes[i] for i in downregulated_idx] # get gene labels\n",
    "    upregulated = [genes[i] for i in upregulated_idx] # get gene labels\n",
    "    \n",
    "    # render l1norm values\n",
    "    x = np.arange(len(q))\n",
    "    y = q\n",
    "    plt.scatter(x, y, s=.1, alpha=1)\n",
    "    plt.axhline(y=cutoff, color='red', linewidth=.5, label='Cutoff')\n",
    "    plt.axhline(y=-cutoff, color='red', linewidth=.5)\n",
    "    plt.xticks([])\t\n",
    "    plt.ylabel('l1-norm')\n",
    "    plt.xlabel('Genes')\n",
    "    plt.legend()\n",
    "    \n",
    "    return downregulated, upregulated\n",
    "\n",
    "# Differential expression between two sparse matrices\n",
    "downregulated, upregulated = diffexp_two_matrices(M1, M2, genes, cutoff=.9) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def subpopulations_grid(pop, method='tsne', figsize=(20,20), size_background=.1, size_subpops=1):\n",
    "'''\n",
    "Generate grid plots of sample subpopulations in an embedding space\n",
    "\n",
    "pop : dict\n",
    "    Popalign object\n",
    "method : str\n",
    "    Embedding method. One of tsne, umap\n",
    "figsize : tuple\n",
    "    Size of the figures to be generated\n",
    "size_background : float, int\n",
    "    Point size for the embedding scatter in the background\n",
    "size_subpops : float, int\n",
    "    Point size for the highlighted subpopulations\n",
    "'''\n",
    "\n",
    "method='tsne'\n",
    "figsize=(20,20)\n",
    "size_background=.1\n",
    "size_subpops=1\n",
    "\n",
    "if method not in pop: # if method not run before\n",
    "    X = cat_data(pop, 'C') # retrieve feature space data\n",
    "    if method == 'umap': # if method is umap\n",
    "        X = umap.UMAP().fit_transform(X) # run umap\n",
    "    elif method == 'tsne': # if method is tsne\n",
    "        X = TSNE(n_components=2).fit_transform(X) # run tsne\n",
    "    else: # if method not valid\n",
    "        raise Exception('Method value not supported. Must be one of tsne, umap.') # raise exception\n",
    "    pop[method] = X # store embedded coordinates\n",
    "else: # if method has been run before\n",
    "    X = pop[method] # retrieve embedded coordinates\n",
    "\n",
    "x = X[:,0] # get x coordinates\n",
    "y = X[:,1] # get y coordinates\n",
    "\n",
    "\n",
    "C = PA.cat_data(pop, 'C') # get sample feature data\n",
    "gmm = pop['gmm'] # get sample gmm\n",
    "poplabels = pop['gmm_types'] # get subpopulations labels\n",
    "prediction = gmm.predict(C) # get subpopulation assignments for the cells\n",
    "\n",
    "n = C.shape[0] # get number of cells\n",
    "nr, nc = nr_nc(gmm.n_components) # get number of rows and columns for the grid plot\n",
    "fig, axes = plt.subplots(nr,nc,figsize=figsize) # create figure and subaxes\n",
    "axes = axes.flatten()\n",
    "\n",
    "for i in range(gmm.n_components): # for each subpopulation of sample\n",
    "    ax = axes[i] # assign sub axis\n",
    "    idx = np.where(prediction==i)[0] # get cell indices for that subpopulations\n",
    "    xtmp = x[idx] # subset cells\n",
    "    ytmp = y[idx] # subset cells\n",
    "    ax.scatter(x, y, c='lightgrey', s=size_background) # plot all cells as background\n",
    "    ax.scatter(xtmp, ytmp, c='purple', s=size_subpops) # plot subpopulation cells on top\n",
    "    ax.set(xticks=[]) # remove x ticks\n",
    "    ax.set(yticks=[]) # remove y ticks\n",
    "    ax.set(title='Subpopulation #%d\\n%s' % (i, poplabels[i])) # set title\n",
    "    if i % nc == 0:\n",
    "        ax.set(ylabel='%s2' % method) # set y label\n",
    "    if i >= len(pop['order'])-nc:\n",
    "        ax.set(xlabel='%s1' % method) # set x label\n",
    "\n",
    "rr = len(axes)-gmm.n_components # count how many empty plots in grid\n",
    "for i in range(1,rr+1):\n",
    "    ax = axes[-i] # backtrack extra sub axes\n",
    "    ax.axis('off') # clear empty axis from plot\n",
    "\n",
    "dname = 'embedding/subpopulations/' # folder name\n",
    "mkdir(os.path.join(pop['output'], dname)) # create folder if does not exist\n",
    "plt.savefig(os.path.join(pop['output'], dname, '%s_subpopulations.png' % sample), dpi=200) # save plot\n",
    "plt.close() # close plot\n",
    "start = end # update start index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3",
   "language": "python",
   "name": "py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
